fl4health/clients/fenda_client.py

```diff
@@ -1,12 +1,220 @@
-from flwr.common.typing import Config
+import copy
+from pathlib import Path
+from typing import Dict, List, Optional, Sequence, Tuple
 
+import torch
+from flwr.common.typing import Config, NDArrays
+
+from fl4health.checkpointing.checkpointer import TorchCheckpointer
 from fl4health.clients.basic_client import BasicClient
 from fl4health.model_bases.fenda_base import FendaModel
 from fl4health.parameter_exchange.layer_exchanger import FixedLayerExchanger
 from fl4health.parameter_exchange.parameter_exchanger_base import ParameterExchanger
+from fl4health.utils.losses import Losses, LossMeterType
+from fl4health.utils.metrics import Metric, MetricMeterType
 
 
 class FendaClient(BasicClient):
+    def __init__(
+        self,
+        data_path: Path,
+        metrics: Sequence[Metric],
+        device: torch.device,
+        loss_meter_type: LossMeterType = LossMeterType.AVERAGE,
+        metric_meter_type: MetricMeterType = MetricMeterType.AVERAGE,
+        checkpointer: Optional[TorchCheckpointer] = None,
+        temperature: float = 0.5,
+        perFCL_loss_weights: List[float] = [0.0, 0.0],
+        cos_sim_loss_weight: float = 0.0,
+        contrastive_loss_weight: float = 0.0,
+    ) -> None:
+        super().__init__(
+            data_path=data_path,
+            metrics=metrics,
+            device=device,
+            loss_meter_type=loss_meter_type,
+            metric_meter_type=metric_meter_type,
+            checkpointer=checkpointer,
+        )
+        self.perFCL_loss_weights = perFCL_loss_weights
+        self.cos_sim_loss_weight = cos_sim_loss_weight
+        self.contrastive_loss_weight = contrastive_loss_weight
+        self.cos_sim = torch.nn.CosineSimilarity(dim=-1)
+        self.ce_criterion = torch.nn.CrossEntropyLoss().to(self.device)
```

Comment:
Do we need to move this criterion to device? 

On a related note, this does highlight something that might be worth folding into this PR, but in the `setup_client` of `BasicClient` when we call `self.get_criterion(config)` should we actually call `self.get_criterion(config).to(self.device)` by default? 
@jewelltaylor: What do you think?

Reply:
In my experience, I think generally it is good to bring all losses on device, however I think it would not particularly change anything from cross entropy loss here, so I deleted it for now.

Reply:
I think that makes a lot of sense. With that in mind. I'm happy for us to just do that by default (i.e. self.cos_sim and self.ce_criterion to device) and also change the BasicClient setup to perform `self.get_criterion(config).to(self.device)` in the `setup_client`